app/src/mouse/input_processor_temp_layer.c

```diff
@@ -9,52 +9,210 @@
+    const struct zmk_position_state_changed *ev = as_zmk_position_state_changed(eh);
+    if (!ev->state) {
+        return ZMK_EV_EVENT_BUBBLE;
+    }
+
+    const struct device *dev = DEVICE_DT_INST_GET(0);
+    struct temp_layer_data *data = (struct temp_layer_data *)dev->data;
+    const struct temp_layer_config *cfg = dev->config;
+
+    if (data->state.is_active && cfg->excluded_positions && cfg->num_positions > 0) {
+        if (!position_is_excluded(cfg, ev->position)) {
+            LOG_DBG("Position not excluded, deactivating layer");
+            update_layer_state(&data->state, false);
+        }
+    }
+    LOG_DBG("Position excluded, continuing");
+
+    return ZMK_EV_EVENT_BUBBLE;
+}
+
+static int handle_keycode_state_changed(const zmk_event_t *eh) {
+    const struct zmk_keycode_state_changed *ev = as_zmk_keycode_state_changed(eh);
+    if (!ev->state) {
+        return ZMK_EV_EVENT_BUBBLE;
+    }
+
+    const struct device *dev = DEVICE_DT_INST_GET(0);
+    struct temp_layer_data *data = (struct temp_layer_data *)dev->data;
+    LOG_DBG("Setting last_tapped_timestamp to: %d", ev->timestamp);
+    data->state.last_tapped_timestamp = ev->timestamp;
+
+    return ZMK_EV_EVENT_BUBBLE;
+}
+
+static int handle_state_changed_dispatcher(const zmk_event_t *eh) {
+    if (as_zmk_position_state_changed(eh) != NULL) {
+        LOG_DBG("Dispatching handle_position_state_changed");
+        return handle_position_state_changed(eh);
+    } else if (as_zmk_keycode_state_changed(eh) != NULL) {
+        LOG_DBG("Dispatching handle_keycode_state_changed");
+        return handle_keycode_state_changed(eh);
+    }
+
+    return ZMK_EV_EVENT_BUBBLE;
+}
+
+/* Driver Implementation */
+static int temp_layer_handle_event(const struct device *dev, struct input_event *event,
+                                   uint32_t param1, uint32_t param2,
+                                   struct zmk_input_processor_state *state) {
+    if (param1 >= MAX_LAYERS) {
+        LOG_ERR("Invalid layer index: %d", param1);
         return -EINVAL;
     }
 
-    if (!zmk_keymap_layer_active(param1)) {
-        zmk_keymap_layer_activate(param1);
+    struct temp_layer_data *data = (struct temp_layer_data *)dev->data;
+    const struct temp_layer_config *cfg = dev->config;
+
+    data->state.toggle_layer = param1;
+
+    if (!data->state.is_active &&
+        !should_quick_tap(cfg, data->state.last_tapped_timestamp, k_uptime_get())) {
+        update_layer_state(&data->state, true);
+    }
+
+    if (param2 > 0) {
+        k_work_reschedule(&layer_disable_works[param1], K_MSEC(param2));
     }
 
-    k_work_reschedule(&layer_disable_works[param1], K_MSEC(param2));
     return 0;
 }
 
-static struct zmk_input_processor_driver_api tl_driver_api = {
-    .handle_event = tl_handle_event,
-};
+static int temp_layer_init(const struct device *dev) {
+    struct temp_layer_data *data = dev->data;
+    data->dev = dev;
+    data->state = (struct temp_layer_state){0};
```

Comment:
This seems superfluous. Is this actually necessary? I can't imagine it is.

Reply:
you're right, removed